allennlp/training/trainer.py

```diff
@@ -769,12 +768,37 @@ def train(self) -> Dict[str, Any]:
                 # No validation set, so just assume it's the best so far.
                 is_best_so_far = True
                 val_metrics = {}
-                best_epoch_val_metrics = {}
                 this_epoch_val_metric = None
 
             self._metrics_to_tensorboard(epoch, train_metrics, val_metrics=val_metrics)
             self._metrics_to_console(train_metrics, val_metrics)
 
+            # Create overall metrics dict
+            training_elapsed_time = time.time() - training_start_time
+            metrics["training_duration"] = time.strftime("%H:%M:%S", time.gmtime(training_elapsed_time))
+            metrics["training_start_epoch"] = epoch_counter,
+            metrics["training_epochs"] = epochs_trained
+            metrics["epoch"] = epoch
+
+            for key, value in train_metrics.items():
+                metrics["training_" + key] = value
+            for key, value in val_metrics.items():
+                metrics["validation_" + key] = value
+
+            if is_best_so_far:
```

Comment:
this shouldn't be behind an if statement, we should always have these otherwise the keys in the dictionary are not constant which is very annoying for anyone trying to use them programmatically.

Reply:
`metrics` is the same object every epoch, so the keys should be the same. (if you don't hit this `if` you keep the "best" values from the previous epoch, and the first epoch you'll for sure hit this if)

(I thought about that there's a weird edge case where if you have different metrics in different epochs funny things could happen, but I decided not to worry about that)